src/main/java/com/bookstore/service/UserServiceImpl.java

```diff
@@ -0,0 +1,38 @@
+package com.bookstore.service;
+
+import com.bookstore.dto.user.UserRegistrationRequestDto;
+import com.bookstore.dto.user.UserResponseDto;
+import com.bookstore.exception.RegistrationException;
+import com.bookstore.mapper.UserMapper;
+import com.bookstore.model.Role;
+import com.bookstore.model.RoleName;
+import com.bookstore.model.User;
+import com.bookstore.repository.user.UserRepository;
+import lombok.RequiredArgsConstructor;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.stereotype.Service;
+import java.util.HashSet;
+import java.util.Set;
+
+@Service
+@RequiredArgsConstructor
+public class UserServiceImpl implements UserService {
+    private final UserRepository userRepository;
+    private final PasswordEncoder passwordEncoder;
+    private final UserMapper userMapper;
+    private final RoleService roleService;
+
+    @Override
+    public UserResponseDto register(UserRegistrationRequestDto request)
+            throws RegistrationException {
+        if (userRepository.findByEmail(request.getEmail()).isPresent()) {
+            throw new RegistrationException(
+                    "The email you have chosen is already in use for another account.");
+        }
+        User user = userMapper.toModel(request);
+        user.setPassword(passwordEncoder.encode(request.getPassword()));
+        Role userRole = roleService.getRoleByRoleName(RoleName.ROLE_USER);
+        user.setRoles(new HashSet<>(Set.of(userRole)));
```

Comment:
```suggestion
        user.setRoles(Set.of(userRole));
```
if you don't intend to modify it later